### PR TITLE
Add chronos2 encoder as another encoder option in OpenTSLMFlamingo

### PR DESCRIPTION
**Pull request by liu-jc**
# Add chronos2 encoder in OpenTSLMFlamingo

## :recycle: Current situation & Problem
The previous encoder only use CNNTokenizer, which I believe that it's patchTST style encoder and training from scratch. This PR mainly tried to add chronos2 as the encoder and we can loaded the pretrained chronos2 to utilize their ability to better understand time series. 

## :gear: Release Notes
* Chronos2Encoder in src/opentslm/model/encoder/
* Add some wandb funcationalities in curricum_learning.py to faciliate the training log

## :books: Documentation
I use in-line comments for describing the implementation, which should be clear. 

## :white_check_mark: Testing
Test file in test/test_chronos2_encoder.py 

### Code of Conduct & Contributing Guidelines
By creating and submitting this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordBDHG/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordBDHG/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by liu-jc** [comment]
Hi team @RealLast @max-rosenblattl @masquare ,
As I talked to Patrick, this implementation is adding the pretrained chronos as the encoder in OpenTSLM. The preliminary result shows good.

**Comment by RealLast** [comment]
Thanks @liu-jc for adding this! Code looks great, just have two minor comments we may want to think about @ThomasKaar  @masquare:

1. Not sure if we should add wandb directly in curriculum_learning; we might want to have a helper class or function to enable it externally, to not make curriculum_learning even longer.
2. We should find a better way to modularize encoders, to avoid having to parse a statement like if encoder == "chronos2" for all possible encoders. We could just pass an encoder via the initializer, e.g. model = OpenTSLMFlamingo(.. encoder=Chronos2()).

**Comment by liu-jc** [comment]
> Thanks @liu-jc for adding this! Code looks great, just have two minor comments we may want to think about @ThomasKaar @masquare:
> 
> 1. Not sure if we should add wandb directly in curriculum_learning; we might want to have a helper class or function to enable it externally, to not make curriculum_learning even longer.
> 2. We should find a better way to modularize encoders, to avoid having to parse a statement like if encoder == "chronos2" for all possible encoders. We could just pass an encoder via the initializer, e.g. model = OpenTSLMFlamingo(.. encoder=Chronos2()).

Thanks @RealLast for the comments. As we discussed, I will focus more on modularize wandb with a helper class and avoid making curriculum_learning.py even longer. 

For modularizing and adding encoders, we maybe delay to another PR later. For this one, we just simply add chronos2encoder. I think the current structure is not too bad. We have the class - "encoder/Chronos2Encoder.py
". which is inherited from TimeSeriesEncoderBase. The only bad thing is that we have to use some statements like ` encoder == "chronos2"`during inference. For this, I think probably we need to modify the inference code and align with the modularization part.

File: src/opentslm/model/llm/OpenTSLMFlamingo.py [hunk 1]
```diff
@@ -145,6 +184,8 @@ def _infer_decoder_layers_attr_name(model):
             model.lang_encoder.get_input_embeddings().requires_grad_(True)
             # TODO: investigate also training the output embeddings when untied
 
+        # print(type(model.vision_encoder), model.vision_encoder)
+        # print(dir(model.vision_encoder))
```
Comment (ThomasKaar) [inline src/opentslm/model/llm/OpenTSLMFlamingo.py]:
```suggestion
```

File: src/opentslm/model/llm/OpenTSLMFlamingo.py [hunk 2]
```diff
@@ -43,18 +54,46 @@ def __init__(
         cross_attn_every_n_layers: int = 1,
         decoder_layers_attr_name: str = None,
         freeze_lm_embeddings: bool = False,
+        encoder_type: str = "chronos2",  # "cnn" or "chronos2"
```
Comment (ThomasKaar) [inline src/opentslm/model/llm/OpenTSLMFlamingo.py:57]: iiuc this will now change the behavior to chronos by default. this is not ideal since we will need change the behavior of `curicullum_training.py` and kind of hides the fact that we just chronos.

i would propose another solution: see comment above.

**Review by ThomasKaar — COMMENTED** [review]
@liu-jc thanks a lot for the amazing work here!!

given also the comment by @RealLast i would propose a slight different approach that keeps the "original" functionality of the paper intact:

let's create 2 new classes.
* `CurriculumTrainierPretrainedEncoder`: inherits from `CurriculumTrainer` and just extends the constructor
* `OpenTSLMFlamingoPretrainedEncoder`: inherits from `OpenTSLMFlamingo` and extends the constructor (potentially also just by hardcoding chronos for now but here we could change the encoder pretty dynamically)

and add a new padding functionality to `pad_and_apply_batch` so we can just naively can call `pad_and_apply_batch(..., padding = "left") and `_initialize_model` add a new OpenTSLMFlamingoPretrainedEncoder-logic.

that way we might be able to keep additional effort for this PR really low but still make sure it is fully backwards compatible and extends the current framework.

wdyt @RealLast @liu-jc @masquare @max-rosenblattl

File: src/opentslm/model/llm/OpenTSLMFlamingo.py (hunk 2 above)
Comment (liu-jc) [reply src/opentslm/model/llm/OpenTSLMFlamingo.py:57]: Hi @ThomasKaar, 
Thanks for the comment! I agree that we may not want to change the default behavior of `curicullum_training.py`. I can change this back. 
But for `CurriculumTrainierPretrainedEncoder`, I feel that would be somewhat similar to the current implementation just another layer. 
For adding a new padding functionality, I agree that it would be great to add `padding` arg.
Will work on this after ICML ddl.

File: curriculum_learning.py [hunk 3]
```diff
@@ -66,6 +69,7 @@
     "stage5_ecg_cot",
 ]
 
+load_dotenv()
```
Comment (masquare) [inline curriculum_learning.py:72]: Can we add an `.env.example` file to the repo, so users know which environment variables the script expects? and then please add `.env` to the .gitignore.

File: curriculum_learning.py [hunk 4]
```diff
@@ -145,7 +156,41 @@ def __init__(
...
+            wandb_project=self.wandb_project,
+            disabled=self.disable_wandb,
+        )
+
+        self.base_dir = os.getenv("AMLT_BLOB_ROOT_DIR")
+        if self.base_dir is None:
+            print('AMLT_BLOB_ROOT_DIR is not set, likely we are using development mode, using current directory')
+            self.base_dir = os.getcwd()
+        else:
+            self.base_dir = os.path.join(self.base_dir, "juncheng","OpenTSLM")
```
Comment (masquare) [inline curriculum_learning.py]: please remove personal paths

File: pyproject.toml [hunk 5]
```diff
@@ -55,6 +55,7 @@ dependencies = [
     "einops>=0.6",
     "wfdb>=4.0",
     "open-flamingo>=0.0.2",
+    "chronos-forecasting>=2.2.0"
```
Comment (masquare) [inline pyproject.toml:58]: if you update the dependencies, can you please also update the lockfile using `uv lock` or `uv sync`

**Review by masquare — COMMENTED** [review]
Thank you for the great work @liu-jc! Please see my comments below.

**Comment by liu-jc** [comment]
Hi @masquare @ThomasKaar @RealLast, 

I updated the code. Basically, removed the personal path, example for .env to support wandb, and added the padding_side.
